### ISSUE TITLE
Differing behavior between Pytorch and dfdx

### ISSUE THREAD

**Issue by opfromthestart** [coreylowman/dfdx#848]
I have made a working model in pytorch as a test and have tried to replicate it using dfdx.
Here are the important parts:
rust model: https://github.com/opfromthestart/touhou-ai/blob/master/src/net/dfdx.rs#L116
python model: https://github.com/opfromthestart/touhou-ai/blob/master/main.py#L99
rust optimizer: https://github.com/opfromthestart/touhou-ai/blob/master/src/net/dfdx.rs#L425
python optimizer: https://github.com/opfromthestart/touhou-ai/blob/master/main.py#L146
However the python version converges well after a couple iterations, while the dfdx version has a weird tiling effect.
The python program trained 2 epochs each on 2000 images, mse 0.0138:
in: ![pyinput0](https://github.com/coreylowman/dfdx/assets/13816619/f3bb2af5-abef-4b5a-9bca-8ed56709beea)
out: ![pyoutput0](https://github.com/coreylowman/dfdx/assets/13816619/1d0e1fee-8974-446e-9769-d59f1bde61bc)
the rust program trained 2 epochs each on 2000 images, mse 0.0627 (i tried many times and could not get better than this):
in: ![as_input](https://github.com/coreylowman/dfdx/assets/13816619/b4273d77-1ac2-4331-86d6-70d2cce000f9)
out: ![encoded](https://github.com/coreylowman/dfdx/assets/13816619/7dc4ac1a-c475-4b32-98c0-bcf5a4862dad)
I may be forgetting something important but it may be somewhere that needs docs. They both use the same optimizer with the same settings, and as far as I know have the same model layout. Some help to conserve parity would be appreciated.

You can test it by cloning the repo and looking in the `dbg_images` folder after running both `python main.py` and `cargo run --release train-encode`.

**Comment by opfromthestart** [comment]
I have confirmed that both use the same image encoding scheme both as input to the images and as output, but the output for the rust one seems to be tending towards negative infinity before the sigmoid (the more epochs it ran, the worse it did).

**Comment by opfromthestart** [comment]
Upon removing the sigmoid layer at the end, the model does train (although still weirdly). What would make pytorch's sigmoid behave well around 0 and 1 that dfdx doesn't do? Also the resulting image has a strange tiling effect that is not present in pytorch.
in: ![as_input](https://github.com/coreylowman/dfdx/assets/13816619/f2cfc083-6957-43a2-a54f-30ae89c75a9f)
out: ![encoded](https://github.com/coreylowman/dfdx/assets/13816619/75a27795-8bbf-4493-91b5-3d73d22b8592)
This is only after 2 epochs of 100 but it shows the tiling behavior.

**Comment by opfromthestart** [comment]
I have tried to replicate the behavior in a small test and I am having no luck, it must be something on my end.

**Comment by opfromthestart** [comment]
I was saving images as (640,400) when it needed to be (400,640).

**Comment by ViliamVadocz** [comment]
@opfromthestart I'm curious. Was there a difference in how you needed to save the images between pytorch and dfdx?

**Comment by opfromthestart** [comment]
@ViliamVadocz I wrote the code to convert images into tensors manually in rust, while I used `torchvision.utils.save_image` in the python version. The python version loaded the images as tensors by themselves so it got the dimensions right by default.
